File: src/common/conversations.py
"""Shared conversation handling framework for Telegram bots."""

import logging
from typing import Callable, Optional, Any, Dict
from telegram import Update
from telegram.ext import ContextTypes

from .models import ConversationFlow, ConversationField

logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)

# Internal state codes
FLOW_COMPLETE = 999  # Sentinel value to signal flow completion


class ConversationContext:
    """Manages conversation state and user data."""
    
    CURRENT_FLOW = "current_flow"
    CURRENT_STEP = "current_step"
    FLOW_DATA = "flow_data"

    @staticmethod
    def set_flow(context: ContextTypes.DEFAULT_TYPE, flow_name: str, flow: ConversationFlow) -> None:
        """Set the current conversation flow."""
        logger.debug(f"🔄 STATE: Setting flow to '{flow_name}' with {flow.step_count()} steps")
        context.user_data[ConversationContext.CURRENT_FLOW] = flow_name
        context.user_data[ConversationContext.CURRENT_STEP] = 0
        context.user_data[ConversationContext.FLOW_DATA] = {}

    @staticmethod
    def get_current_flow(user_data: Dict) -> Optional[str]:
        """Get current flow name."""
        return user_data.get(ConversationContext.CURRENT_FLOW)

    @staticmethod
    def get_current_step(user_data: Dict) -> int:
        """Get current step in flow."""
        return user_data.get(ConversationContext.CURRENT_STEP, 0)

    @staticmethod
    def advance_step(context: ContextTypes.DEFAULT_TYPE) -> None:
        """Move to next step."""
        current_step = context.user_data.get(ConversationContext.CURRENT_STEP, 0)
        next_step = current_step + 1
        context.user_data[ConversationContext.CURRENT_STEP] = next_step
        logger.debug(f"➡️  STATE: Advanced step: {current_step} → {next_step}")

    @staticmethod
    def get_flow_data(user_data: Dict) -> Dict[str, Any]:
        """Get accumulated flow data."""
        return user_data.get(ConversationContext.FLOW_DATA, {})

    @staticmethod
    def set_flow_field(context: ContextTypes.DEFAULT_TYPE, key: str, value: Any) -> None:
        """Set a field in the current flow data."""
        if ConversationContext.FLOW_DATA not in context.user_data:
            context.user_data[ConversationContext.FLOW_DATA] = {}
        context.user_data[ConversationContext.FLOW_DATA][key] = value
        logger.debug(f"📝 DATA: Stored field '{key}' = '{value}'")

    @staticmethod
    def clear_flow(context: ContextTypes.DEFAULT_TYPE) -> None:
        """Clear current flow state."""
        context.user_data.pop(ConversationContext.CURRENT_FLOW, None)
        context.user_data.pop(ConversationContext.CURRENT_STEP, None)
        context.user_data.pop(ConversationContext.FLOW_DATA, None)
        logger.debug(f"✅ STATE: Flow cleared")


class GenericConversationHandler:
    """Generic handler for multi-step conversations."""

    def __init__(self, flows: Dict[str, ConversationFlow]):
        """Initialize handler with available flows."""
        self.flows = flows

    async def handle_input(
        self,
        update: Update,
        context: ContextTypes.DEFAULT_TYPE,
        on_completion: Optional[Callable] = None
    ) -> int:
        """
        Generic handler for conversation input.
        
        Args:
            update: Telegram update
            context: Telegram context
            on_completion: Optional callback when flow completes
            
        Returns:
            State code
        """
        user = update.message.from_user
        flow_name = ConversationContext.get_current_flow(context.user_data)
        current_step = ConversationContext.get_current_step(context.user_data)
        
        logger.debug(f"📨 INPUT: User '{user.first_name}' sent: '{update.message.text}'")
        logger.debug(f"📍 STATE: Current flow='{flow_name}', step={current_step}")
        
        if not flow_name or flow_name not in self.flows:
            logger.error(f"❌ ERROR: Invalid flow: {flow_name}")
            await update.message.reply_text(
                "❌ No active flow. Please use /start to begin."
            )
            return 1
        
        flow = self.flows[flow_name]
        step = flow.get_step(current_step)
        
        if not step:
            logger.error(f"❌ ERROR: Invalid step {current_step} in flow {flow_name}")
            await update.message.reply_text(
                "❌ Error processing your input. Please try again with /start."
            )
            return 1
        
        logger.debug(f"🎯 VALIDATION: Validating input for field '{step.key}'")
        
        # Validate input
        is_valid, error_msg = step.form_field.validate(update.message.text)
        
        if not is_valid:
            logger.warning(f"❌ VALIDATION FAILED: {error_msg}")
            await update.message.reply_text(f"❌ {error_msg}\n\n{step.form_field.prompt}")
            return 1
        
        logger.info(f"✅ VALIDATION PASSED: Field '{step.key}' accepted value: '{update.message.text}'")
        
        # Store the value
        ConversationContext.set_flow_field(context, step.key, update.message.text)
        
        # Check if flow is complete
        total_steps = flow.step_count()
        if current_step + 1 >= total_steps:
            logger.info(f"🎉 COMPLETION: Flow '{flow_name}' is complete!")
            
            flow_data = ConversationContext.get_flow_data(context.user_data)
            logger.debug(f"📦 FLOW_DATA: {flow_data}")
            
            if on_completion:
                logger.debug(f"📞 CALLBACK: Calling on_completion callback")
                await on_completion(update, context, flow_data)
            
            ConversationContext.clear_flow(context)
            logger.debug(f"📤 RESULT: Returning FLOW_COMPLETE signal")
            return FLOW_COMPLETE
        
        # Move to next step
        ConversationContext.advance_step(context)
        next_step_idx = ConversationContext.get_current_step(context.user_data)
        next_step = flow.get_step(next_step_idx)
        
        logger.debug(f"➡️  NEXT_STEP: Prompting for field '{next_step.key}' ({next_step_idx + 1}/{total_steps})")
        
        if next_step:
            logger.info(f"📤 SENDING_PROMPT: About to send prompt for field '{next_step.key}'")
            await update.message.reply_text(next_step.form_field.prompt)
            logger.info(f"✅ PROMPT_SENT: Prompt sent successfully for '{next_step.key}'")
        else:
            logger.error(f"❌ ERROR: Next step {next_step_idx} not found")
            await update.message.reply_text("❌ Error with form. Please restart with /start.")
        
        logger.debug(f"📤 RESULT: Returning ACTIVE_FLOW (continue)")
        return 1

    async def start_flow(
        self,
        update: Update,
        context: ContextTypes.DEFAULT_TYPE,
        flow_name: str,
        target_state: int = 1
    ) -> int:
        """
        Start a new conversation flow.
        
        Args:
            update: Telegram update
            context: Telegram context
            flow_name: Name of the flow to start
            target_state: State to return after flow starts
            
        Returns:
            State code
        """
        logger.debug(f"🚀 FLOW_START: Initiating flow '{flow_name}'")
        
        if flow_name not in self.flows:
            logger.error(f"❌ ERROR: Unknown flow: {flow_name}")
            await update.message.reply_text(
                f"❌ Flow '{flow_name}' not found."
            )
            return target_state
        
        flow = self.flows[flow_name]
        ConversationContext.set_flow(context, flow_name, flow)
        
        logger.info(f"✅ FLOW_STARTED: Flow '{flow_name}' started successfully")
        
        await update.message.reply_text(flow.welcome_message)
        
        first_step = flow.get_step(0)
        if first_step:
            logger.debug(f"📤 FIRST_PROMPT: Sending first prompt for '{first_step.key}'")
            await update.message.reply_text(first_step.form_field.prompt)
        
        logger.debug(f"📤 RESULT: Returning target state {target_state}")
        return target_state

File: workers/expense-bot/src/conversations.js
/**
 * Conversation handling framework for Telegram bots
 * Migrated from Python to JavaScript for Cloudflare Workers
 */

import { ConversationFlow, ConversationField } from './models.js';

// Internal state codes
export const FLOW_COMPLETE = 999; // Sentinel value to signal flow completion

/**
 * Conversation context management for KV storage
 */
export class ConversationContext {
    static CURRENT_FLOW = "current_flow";
    static CURRENT_STEP = "current_step";
    static FLOW_DATA = "flow_data";

    /**
     * Set the current conversation flow
     * @param {KVNamespace} kv - Cloudflare KV namespace
     * @param {string} userId - User ID
     * @param {string} flowName - Name of the flow
     * @param {ConversationFlow} flow - Flow object
     * @returns {Promise<void>}
     */
    static async setFlow(kv, userId, flowName, flow) {
        console.debug(`🔄 STATE: Setting flow to '${flowName}' with ${flow.stepCount()} steps`);
        
        const context = {
            currentFlow: flowName,
            currentStep: 0,
            flowData: {}
        };
        
        await kv.put(`${userId}:context`, JSON.stringify(context));
    }

    /**
     * Get current flow name
     * @param {KVNamespace} kv - Cloudflare KV namespace
     * @param {string} userId - User ID
     * @returns {Promise<string|null>} Current flow name or null
     */
    static async getCurrentFlow(kv, userId) {
        const context = await this.getContext(kv, userId);
        return context?.currentFlow || null;
    }

    /**
     * Get current step in flow
     * @param {KVNamespace} kv - Cloudflare KV namespace
     * @param {string} userId - User ID
     * @returns {Promise<number>} Current step index
     */
    static async getCurrentStep(kv, userId) {
        const context = await this.getContext(kv, userId);
        return context?.currentStep || 0;
    }

    /**
     * Move to next step
     * @param {KVNamespace} kv - Cloudflare KV namespace
     * @param {string} userId - User ID
     * @returns {Promise<void>}
     */
    static async advanceStep(kv, userId) {
        const context = await this.getContext(kv, userId);
        const currentStep = context?.currentStep || 0;
        const nextStep = currentStep + 1;
        
        context.currentStep = nextStep;
        await kv.put(`${userId}:context`, JSON.stringify(context));
        
        console.debug(`➡️  STATE: Advanced step: ${currentStep} → ${nextStep}`);
    }

    /**
     * Get accumulated flow data
     * @param {KVNamespace} kv - Cloudflare KV namespace
     * @param {string} userId - User ID
     * @returns {Promise<Object>} Flow data object
     */
    static async getFlowData(kv, userId) {
        const context = await this.getContext(kv, userId);
        return context?.flowData || {};
    }

    /**
     * Set a field in the current flow data
     * @param {KVNamespace} kv - Cloudflare KV namespace
     * @param {string} userId - User ID
     * @param {string} key - Field key
     * @param {any} value - Field value
     * @returns {Promise<void>}
     */
    static async setFlowField(kv, userId, key, value) {
        const context = await this.getContext(kv, userId);
        if (!context.flowData) {
            context.flowData = {};
        }
        context.flowData[key] = value;
        
        await kv.put(`${userId}:context`, JSON.stringify(context));
        console.debug(`📝 DATA: Stored field '${key}' = '${value}'`);
    }

    /**
     * Clear current flow state
     * @param {KVNamespace} kv - Cloudflare KV namespace
     * @param {string} userId - User ID
     * @returns {Promise<void>}
     */
    static async clearFlow(kv, userId) {
        await kv.delete(`${userId}:context`);
        console.debug(`✅ STATE: Flow cleared`);
    }

    /**
     * Get full context object
     * @param {KVNamespace} kv - Cloudflare KV namespace
     * @param {string} userId - User ID
     * @returns {Promise<Object>} Context object
     */
    static async getContext(kv, userId) {
        try {
            const data = await kv.get(`${userId}:context`, "json");
            return data || {};
        } catch (error) {
            console.error(`❌ ERROR: Failed to get context for ${userId}: ${error.message}`);
            return {};
        }
    }
}

/**
 * Generic handler for multi-step conversations
 */
export class GenericConversationHandler {
    /**
     * Initialize handler with available flows
     * @param {Object} flows - Object with flow names as keys and ConversationFlow objects as values
     */
    constructor(flows) {
        this.flows = flows;
    }

    /**
     * Generic handler for conversation input
     * @param {KVNamespace} kv - Cloudflare KV namespace
     * @param {string} userId - User ID
     * @param {string} input - User input text
     * @param {Function} onCompletion - Optional callback when flow completes
     * @returns {Promise<number>} State code
     */
    async handleInput(kv, userId, input, onCompletion = null) {
        try {
            const flowName = await ConversationContext.getCurrentFlow(kv, userId);
            const currentStep = await ConversationContext.getCurrentStep(kv, userId);
            
            console.debug(`📨 INPUT: User '${userId}' sent: '${input}'`);
            console.debug(`📍 STATE: Current flow='${flowName}', step=${currentStep}`);
            
            if (!flowName || !this.flows[flowName]) {
                console.error(`❌ ERROR: Invalid flow: ${flowName}`);
                return 1;
            }
            
            const flow = this.flows[flowName];
            const step = flow.getStep(currentStep);
            
            if (!step) {
                console.error(`❌ ERROR: Invalid step ${currentStep} in flow ${flowName}`);
                return 1;
            }
            
            console.debug(`🎯 VALIDATION: Validating input for field '${step.key}'`);
            
            // Validate input
            const validation = step.formField.validate(input);
            
            if (!validation.isValid) {
                console.warning(`❌ VALIDATION FAILED: ${validation.errorMessage}`);
                return 1;
            }
            
            console.info(`✅ VALIDATION PASSED: Field '${step.key}' accepted value: '${input}'`);
            
            // Store the value
            await ConversationContext.setFlowField(kv, userId, step.key, input);
            
            // Check if flow is complete
            const totalSteps = flow.stepCount();
            if (currentStep + 1 >= totalSteps) {
                console.info(`🎉 COMPLETION: Flow '${flowName}' is complete!`);
                
                const flowData = await ConversationContext.getFlowData(kv, userId);
                console.debug(`📦 FLOW_DATA: ${JSON.stringify(flowData)}`);
                
                if (onCompletion) {
                    console.debug(`📞 CALLBACK: Calling on_completion callback`);
                    await onCompletion(kv, userId, flowData);
                }
                
                await ConversationContext.clearFlow(kv, userId);
                console.debug(`📤 RESULT: Returning FLOW_COMPLETE signal`);
                return FLOW_COMPLETE;
            }
            
            // Move to next step
            await ConversationContext.advanceStep(kv, userId);
            const nextStepIdx = await ConversationContext.getCurrentStep(kv, userId);
            const nextStep = flow.getStep(nextStepIdx);
            
            console.debug(`➡️  NEXT_STEP: Prompting for field '${nextStep.key}' (${nextStepIdx + 1}/${totalSteps})`);
            
            return 1; // Continue flow
        } catch (error) {
            console.error(`❌ ERROR in handleInput: ${error.message}`);
            return 1;
        }
    }

    /**
     * Start a new conversation flow
     * @param {KVNamespace} kv - Cloudflare KV namespace
     * @param {string} userId - User ID
     * @param {string} flowName - Name of the flow to start
     * @param {number} targetState - State to return after flow starts
     * @returns {Promise<number>} State code
     */
    async startFlow(kv, userId, flowName, targetState = 1) {
        console.debug(`🚀 FLOW_START: Initiating flow '${flowName}'`);
        
        if (!this.flows[flowName]) {
            console.error(`❌ ERROR: Unknown flow: ${flowName}`);
            return targetState;
        }
        
        const flow = this.flows[flowName];
        await ConversationContext.setFlow(kv, userId, flowName, flow);
        
        console.info(`✅ FLOW_STARTED: Flow '${flowName}' started successfully`);
        
        return targetState;
    }
}